src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php

```diff
@@ -316,11 +325,18 @@ public function updateAction(Request $request)
                 return $this->redirectHandler->redirectToResource($configuration, $resource);
             }
 
-            if ($configuration->hasStateMachine()) {
-                $this->stateMachine->apply($configuration, $resource);
+            try {
+                $this->resourceUpdater->applyTransitionAndFlush($resource, $configuration, $this->manager);
+            } catch (RaceConditionException $exception) {
+                if (!$configuration->isHtmlRequest()) {
+                    return $this->viewHandler->handle($configuration, View::create($form, Response::HTTP_BAD_REQUEST));
```

Comment:
How about having the code on the exception? We can have `$raceConditionException->getApiResponseCode() // return 423 Locked` and e.g. `$differentException->getApiResponseCode() // return 4xx MyCode`. Sounds like the most generic approach 😄